service/cloud/gcp_streamer_test.go

```diff
@@ -0,0 +1,116 @@
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package cloud_test
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	gcloud "cloud.google.com/go/storage"
+	"github.com/fxamacker/cbor/v2"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/api/option"
+
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/service/cloud"
+	"github.com/optakt/flow-dps/testing/mocks"
+)
+
+func TestGCPStreamer_Next(t *testing.T) {
+	blockData := mocks.GenericBlockData()
+	data, err := cbor.Marshal(blockData)
+	require.NoError(t, err)
+
+	t.Run("nominal case", func(t *testing.T) {
+		server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, _ *http.Request) {
+			rw.WriteHeader(http.StatusOK)
+		}))
+
+		client, err := gcloud.NewClient(
+			context.Background(),
+			option.WithoutAuthentication(),
+			option.WithEndpoint(server.URL),
+		)
+		require.NoError(t, err)
+		bucket := client.Bucket("test")
+
+		buffer := dps.NewDeque()
+		buffer.PushFront(blockData)
+
+		streamer := cloud.BaselineStreamer(t, cloud.WithBucket(bucket), cloud.WithBuffer(buffer))
+
+		got, err := streamer.Next()
+
+		require.NoError(t, err)
+		assert.Equal(t, blockData, got)
+	})
+
+	t.Run("returns unavailable when no block data in buffer", func(t *testing.T) {
+		server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, _ *http.Request) {
+			rw.WriteHeader(http.StatusOK)
+		}))
+
+		client, err := gcloud.NewClient(
+			context.Background(),
+			option.WithoutAuthentication(),
+			option.WithEndpoint(server.URL),
+		)
+		require.NoError(t, err)
+		bucket := client.Bucket("test")
+
+		streamer := cloud.BaselineStreamer(t, cloud.WithBucket(bucket))
+
+		_, err = streamer.Next()
+
+		require.Error(t, err)
+		assert.ErrorIs(t, err, dps.ErrUnavailable)
+	})
+
+	t.Run("downloads records from queue", func(t *testing.T) {
+		serverCalled := make(chan struct{})
+		server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, _ *http.Request) {
+			_, _ = rw.Write(data)
+			serverCalled <- struct{}{}
+		}))
+
+		client, err := gcloud.NewClient(
+			context.Background(),
+			option.WithoutAuthentication(),
+			option.WithEndpoint(server.URL),
+		)
+		require.NoError(t, err)
+		bucket := client.Bucket("test")
+
+		queue := dps.NewDeque()
+		queue.PushFront(blockData.Block.ID())
+
+		streamer := cloud.BaselineStreamer(t, cloud.WithBucket(bucket), cloud.WithQueue(queue))
+
+		_, err = streamer.Next()
+
+		require.Error(t, err)
+		assert.ErrorIs(t, err, dps.ErrUnavailable)
+
+		select {
+		case <-time.After(100 * time.Millisecond):
+			t.Fatal("GCP Streamer did not attempt to download record from bucket")
+		case <-serverCalled:
```

Comment:
Then this should be done in an internal test. I'll move it over so that we can do the assertions you're describing.

Reply:
Unfortunately we can't test that second case because the polling goroutine ends up getting stuck on `reader, err := object.NewReader(context.Background())` for long enough for the test to timeout, so we can't verify that the queue gets filled back after that call returns an error.

What do you suggest? Are you fine with not testing this?